# QCatch
Quality Control downstream of alevin-fry / simpleaf.

**QCatch** is a Python package designed to streamline quality control for single-cell sequencing data quantified by [alevin-fry](https://github.com/COMBINE-lab/alevin-fry) or [simpleaf](https://github.com/COMBINE-lab/simpleaf). It provides a comprehensive web-based quality control report, enabling researchers to:

- Summarize key **quality metrics** for single-cell sequencing datasets.
- Perform **cell calling** to identify high-quality cells.
- Generate interactive **visualizations** to support downstream analysis and interpretation.

QCatch is built to simplify the quality control process, making it easier for researchers to assess data quality and make informed decisions for further analysis.

## Installation

### Bioconda
You can install using [Conda](http://anaconda.org/)
from [Bioconda](https://bioconda.github.io/).

```bash
conda install -c bioconda qcatch 
```

### PyPI

You can also install from [PyPI](https://pypi.org/project/qcatch/) using `pip`:

```bash
pip install qcatch
```

## Basic Usage
Provide the path to the folder containing quantification results (or directly to a .h5ad file) generated by `alevin-fry` or `simpleaf`. QCatch will automatically scan the folder or load the specified file, assess data quality, and generate an interactive HTML report that can be viewed directly in your browser.

```bash
qcatch \
    --input path/to/your/quants/result/folder \
    --output path/to/desired/QC/output/folder \
    --chemistry 10X_3p_v3 
```
We highly recommend specifying the chemistry used in your experiment. By default, QCatch will assume the settings for 10X 3' v2 and v3 chemistry.

## Command-Line Arguments

| Flag | Short | Type | Description |
|------|-------|------|-------------|
| `--input`  | `-i` | `str` (Required) | Path to the input directory containing the quantification output files or to the HDF5 file itself. |
| `--output` | `-o` | `str`(Required)  | Path to the output directory.|
| `--chemistry` | `-c` | `str`(Optional but recommend) | Specifies the chemistry used in the experiment, determining the range for the `empty_drops` step. **Options**: `'10X_3p_v2'`, `'10X_3p_v3'`, `'10X_3p_v4'`, `'10X_3p_LT'`,`'10X_5p_v3'`,`'10X_HT'`. **Default**: Will use the range for `'10X_3p_v2'` and `'10X_3p_v3'`. |
| `--n_partitions` | `-n` | `int` (Optional) | Number of partitions (max number of barcodes to consider for ambient estimation). Skip this step if you already specified `--chemistry`. Only use `--n_partitions` when your experiment uses a custom chemistry not listed in the predefined chemistry options.|
| `--gene_id2name_file` | `-g` | `str` (Optional) | Fail provides a mapping from gene IDs to gene names. The file must be a TSV containing two columns—‘gene_id’ (e.g., ENSG00000284733) and ‘gene_name’ (e.g., OR4F29)—without a header row. If not provided, the program will attempt to retrieve the mapping from a remote registry. If that lookup fails, mitochondria plots will not be displayed.|
| `--save_filtered_h5ad` | `-s` | `flag` (Optional) |If enabled, `qcatch` will save a separate `.h5ad` file containing only the retained cells.|
| `--overwrite_h5ad` | `-w` | `flag` (Optional) |If enabled, `qcatch` will overwrite the original `.h5ad` file in place by appending cell filtering results to `anndata.obs`. No existing data or cells will be removed; only additional metadata columns are added.|
| `--skip_umap_tsne` | `-u` | `flag` (Optional) | If provided, skips generation of UMAP and t-SNE plots. |
| `--verbose` | `-b` | `flag` (Optional) | Enable verbose logging with debug-level messages. |
| `--version` | `-v` | `flag` (Optional) | Display the installed version of qcatch. |
