// @octostar/platform-api/dist/components/OsSpinner.js
/**
 * @beta
 * Spinner component.
 */
export class OsSpinner extends HTMLElement {
    constructor() {
        super();
        this.position = 'floating';
        this.image = '/static/assets/images/loading.svg';
        this.root = this.attachShadow({ mode: 'open' });
        const attrImage = this.getAttribute('image');
        if (attrImage) {
            this.image = attrImage;
        }
        this.render();
    }
    static get observedAttributes() {
        return ['position', 'image'];
    }
    attributeChangedCallback(name, oldValue, newValue) {
        if (oldValue !== newValue) {
            if (name === 'position') {
                this.position = newValue || 'floating';
            }
            if (name === 'image') {
                this.image = newValue || this.image;
            }
            this.render();
        }
    }
    set imageSrc(value) {
        this.image = value || this.image;
        this.setAttribute('image', this.image);
        this.render();
    }
    get imageSrc() {
        return this.image;
    }
    render() {
        this.root.innerHTML = `
    <style>
    .loading {
      z-index: 99;
      width: 50px;
      height: unset;
      position: relative;
      margin: 10px;
    }
    .inline {
      margin: 0px;
      width: 30px;
    }
    .inline-centered {
      margin: 0 auto;
      width: 30px;
      display: block;
    }
    .floating {
      padding: 0;
      margin: 0;
      position: absolute;
      left: 50%;
      top: 50%;
      transform: translate(-50%, -50%);
    }
    </style>
    `;
        const img = document.createElement('img');
        img.src = this.image;
        img.className = `loading ${this.position}`;
        img.alt = 'loading...';
        this.root.innerHTML = '';
        this.root.appendChild(img);
    }
}
if (!customElements.get('os-spinner')) {
    customElements.define('os-spinner', OsSpinner);
}
