// @octostar/platform-types/dist/model.js
/**
 * Enum for sort order values.
 * @beta
 */
export var SortOrder;
(function (SortOrder) {
    SortOrder[SortOrder["PREFERRED"] = 21] = "PREFERRED";
    SortOrder[SortOrder["OTHER"] = 10] = "OTHER";
    SortOrder[SortOrder["HIDE"] = -1] = "HIDE";
})(SortOrder || (SortOrder = {}));
/**
 * Possible statuses of a NiFi pipeline task.
 * @beta
 */
export var ProcessingStatusCode;
(function (ProcessingStatusCode) {
    /** Queued and awaiting execution */
    ProcessingStatusCode["QUEUED"] = "QUEUED";
    /** Currently running */
    ProcessingStatusCode["RUNNING"] = "RUNNING";
    /** Stopped manually */
    ProcessingStatusCode["STOPPED"] = "STOPPED";
    /** Paused and can be resumed */
    ProcessingStatusCode["PAUSED"] = "PAUSED";
    /** Completed successfully */
    ProcessingStatusCode["COMPLETED"] = "COMPLETED";
    /** Encountered a failure */
    ProcessingStatusCode["FAILED"] = "FAILED";
    /** Scheduled to run */
    ProcessingStatusCode["SCHEDULED"] = "SCHEDULED";
    /** Starting up */
    ProcessingStatusCode["STARTING"] = "STARTING";
    /** Shutting down */
    ProcessingStatusCode["STOPPING"] = "STOPPING";
    /** Disabled and cannot run */
    ProcessingStatusCode["DISABLED"] = "DISABLED";
    /** Invalid configuration */
    ProcessingStatusCode["INVALID"] = "INVALID";
    /** Waiting for an external event */
    ProcessingStatusCode["WAITING"] = "WAITING";
    /** Retrying after failure */
    ProcessingStatusCode["RETRYING"] = "RETRYING";
    /** Status is unknown */
    ProcessingStatusCode["UNKNOWN"] = "UNKNOWN";
})(ProcessingStatusCode || (ProcessingStatusCode = {}));
