// @octostar/platform-api/dist/components/OsTagsContextMenu.js
import { AbstractContextMenu } from './AbstractContextMenu';
/**
 * @beta
 * Custom element that renders a context menu for tags.
 */
export class OsTagsContextMenu extends AbstractContextMenu {
    constructor() {
        super(...arguments);
        this.entity = undefined;
    }
    static get observedAttributes() {
        return ['entity'];
    }
    connectedCallback() {
        super.connectedCallback();
        this.setAttribute('slot', 'tags-menu-slot');
        this.render();
    }
    attributeChangedCallback(name, oldValue, newValue) {
        if (name === 'entity' && newValue) {
            this.record = newValue;
        }
    }
    set record(entity) {
        this.entity = entity;
        this.update();
    }
    getContext() {
        return { item: this.entity, menuType: 'tags' };
    }
    render() {
        this.innerHTML = `
      <style>
        .tags-context-menu {
          display: inline-flex;
          align-items: center;
          gap: 8px;
          background: #f0f0f0;
          padding: 4px 8px;
          border-radius: 4px;
        }
        .anticon {
          display: inline-flex;
          justify-content: center;
          align-items: center;
        }
      </style>
    <div class="tags-context-menu">
      <span role="img" aria-label="tag" class="anticon anticon-tag">
      <svg viewBox="64 64 896 896" focusable="false" data-icon="tag" width="1em" height="1em" fill="currentColor" aria-hidden="true">
      <path d="M938 458.8l-29.6-312.6c-1.5-16.2-14.4-29-30.6-30.6L565.2 86h-.4c-3.2 0-5.7 1-7.6 2.9L88.9 557.2a9.96 9.96 0 000 14.1l363.8 363.8c1.9 1.9 4.4 2.9 7.1 2.9s5.2-1 7.1-2.9l468.3-468.3c2-2.1 3-5 2.8-8zM459.7 834.7L189.3 564.3 589 164.6 836 188l23.4 247-399.7 399.7zM680 256c-48.5 0-88 39.5-88 88s39.5 88 88 88 88-39.5 88-88-39.5-88-88-88zm0 120c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32z"></path></svg></span><span role="img" aria-label="plus" class="anticon anticon-plus"><svg viewBox="64 64 896 896" focusable="false" data-icon="plus" width="1em" height="1em" fill="currentColor" aria-hidden="true"><path d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"></path><path d="M192 474h672q8 0 8 8v60q0 8-8 8H160q-8 0-8-8v-60q0-8 8-8z"></path>
      </svg></span></div>`;
    }
    update() {
        this.render();
    }
}
if (!customElements.get('os-tags-context-menu')) {
    customElements.define('os-tags-context-menu', OsTagsContextMenu);
}
